### PR TITLE
ISPN-14782 Unable to reenable rebalance after cluster scale up

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-14782
https://issues.redhat.com/browse/ISPN-14793

We do not broadcast the topology as stable if it wasn't yet restored.

### CODE REVIEW

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#1](https://ci.infinispan.org/job/Infinispan/job/PR-10834/1/):
```
quay.io/infinispan-test/server:PR-10834
```

**Comment by jabolina** [comment]
Replayed CI.

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#2](https://ci.infinispan.org/job/Infinispan/job/PR-10834/2/):
```
quay.io/infinispan-test/server:PR-10834
```

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#3](https://ci.infinispan.org/job/Infinispan/job/PR-10834/3/):
```
quay.io/infinispan-test/server:PR-10834
```

File: core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java [hunk 1]
```diff
@@ -827,8 +827,15 @@ public synchronized void startQueuedRebalance() {
          return;
       }
 
+      // We don't have a queued rebalance.
       if (queuedRebalanceMembers == null) {
-         // We don't have a queued rebalance. We may need to broadcast a stable topology update
+         // The previous topology was not restored. We do nothing, waiting for the members to get back in and install topology.
+         if (currentTopology == null && persistentState.isPresent()) {
+            log.debugf("Skipping rebalance for cache %s as the previous topology was not restored", cacheName);
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java:834]: I'm now getting stuck at this message and all operations on the cache return:

```
ISPN000689: Recovering cache 'persistentCache' but there are missing members, known members [test-upgrade-1-64083] of a total of 2
```
In both sets of logs on the Jira, the cluster receives a MERGED view from JGroups. The reproducers don't include this.

```
test-upgrade-0 infinispan 14:20:09,108 INFO  (jgroups-6,test-upgrade-0-13791) [org.infinispan.CLUSTER] ISPN000093: Received new, MERGED cluster view for channel test-upgrade: MergeView::[test-upgrade-1-60105|3] (2) [test-upgrade-1-60105, test-upgrade-0-13791], 2 subgroups: [test-upgrade-1-60105|1] (2) [test-upgrade-1-60105, test-upgrade-0-13791], [test-upgrade-1-60105|2] (1) [test-upgrade-1-60105]
test-upgrade-1 infinispan 14:20:09,109 INFO  (jgroups-6,test-upgrade-1-60105) [org.infinispan.CLUSTER] ISPN000093: Received new, MERGED cluster view for channel test-upgrade: MergeView::[test-upgrade-1-60105|3] (2) [test-upgrade-1-60105, test-upgrade-0-13791], 2 subgroups: [test-upgrade-1-60105|1] (2) [test-upgrade-1-60105, test-upgrade-0-13791], [test-upgrade-1-60105|2] (1) [test-upgrade-1-60105]
```

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#4](https://ci.infinispan.org/job/Infinispan/job/PR-10834/4/):
```
quay.io/infinispan-test/server:PR-10834
```

**Comment by jabolina** [comment]
Added a commit for https://issues.redhat.com/browse/ISPN-14793

This changes how the nodes handle the `CacheStatusRequest` a bit. This request is sent by the coordinator after a view MERGE. If the node has a cache that needs to be recreated from the persistent state, it will send a join request to the coordinator. The coordinator only handles this request _after_ receiving all the states from the nodes, and, in case we are partitioned again during this, the current join mechanism should kick in.

File: core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java (hunk 1 above)
Comment (jabolina) [reply core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java:834]: We should be covering that with the latest commit :pray: :pray:

**Review by ryanemerson — APPROVED** [review]
LGTM just two minor points.

I've backported to speed things up while you're PTO: https://github.com/infinispan/infinispan/pull/10849

I removed the `System.out` comment, but I haven't changed the `AggregateCompletionStage` usage.

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java [hunk 2]
```diff
@@ -295,6 +309,7 @@ public CompletionStage<ManagerStatusResponse> handleStatusRequest(int viewId) {
       // As long as we have an older view, we can still process topologies from the old coordinator
       return withView(viewId, getGlobalTimeout(), MILLISECONDS).thenApply(ignored -> {
          Map<String, CacheStatusResponse> caches = new HashMap<>();
+         AggregateCompletionStage<Void> joins = CompletionStages.aggregateCompletionStage();
```
[the remark this answers is not in the document]
Comment (jabolina) [reply core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java:312]: Sorry, this was leftover from the first solution. Updating to remove it.

**Comment by jabolina** [comment]
Updated with suggestions.

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#5](https://ci.infinispan.org/job/Infinispan/job/PR-10834/5/):
```
quay.io/infinispan-test/server:PR-10834
```

**Pull request merged**